### ISSUE TITLE
Modelling translation

### ISSUE THREAD

**Issue by kellertuer** [JuliaManifolds/LieGroups.jl#3]
I think this is a discussion for @olivierverdier and @mateuszbaran because I think I narrowed down where I got lost in the remodelling of Manifolds.jl 0.10.0 just now and would like to discuss this before I continue to model translations here

In 0.10.0 we have

* $L_p(q) = pq$ or the usual left (forward?) translation
* $R_p(q) = qp$ or the usual right (forward?) translation.

And we have their inverses

* $L^{-1}_p(q) = p^{-1}q$
* $R^{-1}_p(q) = qp^{-1}$

which are the (forward?) left and right inverse translations, respectively.

But we also have $L'_p(q) = p^{-1}q$ and $R'_p = qp^{-1}$ – (backward) translations? Why? The inverses do exactly the same thing furthermore also have the inverse backwards ones which are just the forward non inverse ones.

Why do we define so many different things that all do the same? Can we just leave out forward backward and use translation and its inverse? I do not understand why these are necessary, to me that all sounds like duplication to confuse users. (If we want to confuse users we could further introduce [widdershins and deosil](https://en.wikipedia.org/wiki/Widdershins) $L^w=R^d$, $L^d=R^w$ also doing the same thing ;) )

**Comment by olivierverdier** [comment]
We need `translate` and `translate_diff`, but the inverse versions should go: `inverse_translate` and `inverse_translate_diff`, as they are redundant now.

**Comment by kellertuer** [comment]
Well, the first good point is, then I know where most of my confusion in the 2-3 PRs for this came from, I was always of the impression we double things.

But great that you agree that we should not double things or provide too many names for the same operation. Do you have a reference for the forward/backward notation choice? Is that the best naming?

**Comment by olivierverdier** [comment]
A deprecation would be nice for users (even in Manifolds.jl):
```julia
@deprecate inverse_translate(G, p, q, (conv, side)) translate(G, p, q, (switch_direction(conv), side))
```

**Comment by olivierverdier** [comment]
Yes, things were duplicated, even worse, currently the code is duplicated, hence the deprecation above.

The name "backward"/"forward" is the first one that came to my mind, and @mateuszbaran used it. It's really worth thinking carefully about a possibly more suitable name.

**Comment by kellertuer** [comment]
Here, since this is a new start, I would only introduce one version.
Since this a completely new package – on purpose among others for – designing things a bit differently. So here I would just not introduce the inverse_translate – but mention it in the News.md.
Maybe one could mention it on a page “Transition from Manifolds.jl” for sure.

What I liked in `inverse_translate` is that we have two nice and different names for them. But I do see that in one function the “switching” between these 4 operations is easier. I would maybe use `Left`, `LeftInverse`, `Right` and `RightInverse`?

**Comment by kellertuer** [comment]
For Manifolds.jl the deprecation would not make so much sense – we would remove groups in 0.11 and until then the deprecation is not so useful I think.

**Comment by olivierverdier** [comment]
What would `RightInverse` be?

**Comment by kellertuer** [comment]
* $R$ is `Right` (currently right forward)
* $R^{-1}$ is `RightInverse` (currently right backward)
* $L$ is `Left` (currently left forward)
* $L^{-1}$ is `LeftInverse` (currently left backward)

this stays closer to the old function names and I feel `X` and `XInverse` are also closer to what we write, remembering forward/backwards would be new terminology to that. At least I thought there is some new magic in there if these need new names (when they were indeed the `inv_translate` ones).

**Comment by olivierverdier** [comment]
It's a good idea, but it doesn't work, as the inverse of right forward is left forward. The inverse of any action in general switches left/right.

**Comment by olivierverdier** [comment]
Other suggestions:
- regular / dual (I use that in one of my papers)
- regular / inverse 
- direct / opposite

None is very satisfactory. I wish there were a nice solution to that naming problem (perhaps there is one?).

**Comment by olivierverdier** [comment]
The root cause of the problem is the tradition to call "covariant" action a left action, and a "contravariant" action a right action. This is a huge mistake.

**Comment by olivierverdier** [comment]
In point of fact, a good name would be
- covariant left for pq
- contravariant left for p^{-}q
- covariant right for q p^{-1}
- contravariant right for q p

And change left/right of all actions to covariant/contravariant.

**Comment by kellertuer** [comment]
Ah I see, I mixed group ops and actions in my head for those names.

Well we have a few ideas here, and we can just leave that open and think a bit (my widdershins/deosil is maybe similar to poleward/rimward, but merely a Pratchet-reference).

Hm now you lost me a bit, it is a huge mistake by mathematicians – but then you propose to use exactly those names?

I think having Left/right (inverse) for group operations and co/contravariant for actions would also nicely distinguish actions and group operations, so for me that sounds good, and of course we would carefully document all these.

**Comment by olivierverdier** [comment]
<warning, category math ahead>
Covariant and contravariant are not that stupid in fact: a left action is precisely a covariant functor from a group to the permutations of a set, and a right action is a covariant functor.
</warning>

**Comment by olivierverdier** [comment]
> but then you propose to use exactly those names?

No! I suggest to drop left/right in favor of covariant/contravariant.

THEN, left and right really mean left and right multiplications.

**Comment by kellertuer** [comment]
So to summarise (maybe for Metausz to agree)

* left: left group operation $f_h(g) = hg$
* right: tight group operation $f_h(g) = gh$

but then afterwards you now lost me completely. The names are totally wrong with a huge mistake but we still use them to drop left/right but you still write

> * covariant left for pq

so we do not drop them. You lost me on both the huge mistake we still use and the drop we do not drop 😄

**Comment by kellertuer** [comment]
(or in other words, maybe make your naming scheme complete: both group operations names and all 4 actions, there was a bit too much back and forth huge error, still using, not using half replacement changes in the last posts for me)

**Comment by olivierverdier** [comment]
Ok, let's do it in two steps:

**Step one** Drop the names left/right *completely*: `LeftAction` -> `CovariantAction`, `RightAction` -> `ContravariantAction`

(and maybe then `RightForward` -> `ContravariantForward`, etc.)

Ok so far?

**Comment by olivierverdier** [comment]
Now, suddently, the names Right/Left are *free*.

So now we realise that the forward action is just multiplications on the left, so we call it left action instead. Which leads to

**step two** Replace `Forward` by `Left` and `Backward` by `Right`

**Comment by olivierverdier** [comment]
So the names of the four actions on a group on itself are now:


 -    covariant left for $q \mapsto pq$
 -   contravariant left for $q \mapsto p^{-}q$
 -  covariant right for $q \mapsto q p^{-1}$
 - contravariant right for $q \mapsto q p$

Now when the name contains *left*, it is a multiplication on the *left*.

**Comment by kellertuer** [comment]
Thanks with the summary I can follow, the derivation of renaming, I could not follow, for 4 things we do I endet up with 12-16 names on my sheet and got lost again.

so, to summarise

* the formerly `Left` then `LeftForward` will become `CovariantLeft`
* the formerly inverse `Left` then `LeftBackward` will become `ContravariantLeft`
* the formerly `Right` then `RightForward` will become `CovariantRight`
* the formerly inverse `Right` then `RightBackward` will become `ContravariantRight`

eh? No, tthat is now what you got and I do still not know how you endet up there! Now I am fully confused and nearly at the point, that I should maybe not program but do something with wood or so.
I have no clue how you endet up with that naming scheme.

So formerly we had a name for anything indicating left or right. That is still visible.
But covariant sometimes is p sometimes its inverse. And that is what for?

I am a bit new to all your names and slowly prefer the 4 names from Pratchet since they make as much sense to me now here as well.
So maybe instead of a wild twist of moving and shuffling names until I am totally 😵‍💫 how about:

We name the 4 things: A, B, C, D because Book X says so?

**Comment by kellertuer** [comment]
or in other words, at some point you moved from the second word referring to p or its inverse to that switching between R and L

The current scheme is (order important) LF, RF, LB, RB, but you now do LCov, RCon, LCon, RCov, so I have no real idea what Con/Cov here refer to, they do the opposite as before F and B did. No clue why, since your derivation was more like a shuffling of a deck of cards with some cards becoming “free”.

**Comment by olivierverdier** [comment]
> so, to summarise

    the formerly Left then LeftForward will become CovariantLeft
    the formerly inverse Left then LeftBackward will become ContravariantLeft
    the formerly Right then RightForward will become CovariantRight
    the formerly inverse Right then RightBackward will become ContravariantRight

Your summary is **incorrect**.

**Comment by olivierverdier** [comment]
I suggest replacing `Left` with `Covariant`, and `Right` with `Contravariant`.

**Comment by olivierverdier** [comment]
So, `LeftForward` -> `CovariantForward`,
`LeftBackward` -> `CovariantBackward`,
`RightForward` -> `ContravariantForward`
`RightBackward` -> `ContravariantBackward`

It's a simple substitution.

**Comment by olivierverdier** [comment]
That was the *first step*.

And now, look at what `Forward` currently means: it means `LeftSide`, and `Backward` means `RightSide`.

**Comment by kellertuer** [comment]
I know that my summary is incorrect. But I only replaced names as you said, so at some point you do not replace but reorder.
And I do not understand where that happens. In the old form one can see (a) left/right (b) inverse/noninverse.
In my try to follow your shuffle that is still present, but it is incorrect since it does not agree with yours.

So somewhere you switch something. But you never say you switch something. I do not know when how or even why. Because the new naming scheme Contra/Covariant does not reflect any property for me.


Oh now we also get sides in there as well. Uff. I give up. Can we just use some naming from a book and not introduce 24 names in some shuffle reorder rename game ?

**Comment by olivierverdier** [comment]
It's two substitutions one after the other, and the order matters.

**Comment by olivierverdier** [comment]
No, `LeftSide` is what *you currently have* as an alias for `Forward`.

**Comment by kellertuer** [comment]
No Forward means: use p. Left means left
Backward means use p^-1, right means right.

And at some point you introduce some new meaning to the second variable you use so that Contra and Covariant mean – some magic I have not yet understood. But at this point this gets infinitely long so I give up.

**Comment by olivierverdier** [comment]
> No Forward means: use p. Left means left
Backward means use p^-1, right means right.

That is not the meaning of backward forward, no.

**Comment by kellertuer** [comment]
Let's do it that way. 

You propose some names after this comment. If Mateusz agrees we take them. I will not understand them anyways, because you now try for over an hour and I do not see a chance. At some point you do some magic flip that I do not follow.

**Comment by kellertuer** [comment]
I can also just be the coding monkey of this package then. That is ok.

**Comment by olivierverdier** [comment]
Ok, here is the final substitutions:

1. Everywhere in the code (*except* Left/RightSide), replace
  - `Left` -> `Covariant`
  - `RIght` -> `Contravariant`
2. Drop the all the alias containing `Forward`/`Backward` as they are just confusing.

**Comment by kellertuer** [comment]
Ah I see in the docs: Left backward does a multiplication from the right. Very intuitive.

Yes I do not follow any of this. As I said, I am okay with just being the stupid coding monkey of this. There was already a reason why I stopped trying to follow all the 0.10.0 rework, both because I felt it was redundant as well as a lot of statement that everything is fundamentally wrong and should definetly be changed, but I never got around to what is wrong.

But again. It is probably time to give up from me trying to understand this then. We take what you propose after this post and that's it. Coding monkey will then code.

**Comment by olivierverdier** [comment]
So you should replace everywhere `LeftAction` by `CovariantAction` and `RightAction` by `ContravariantAction`

- `(LeftAction, LeftSide)`  by `(CovariantAction, LeftSide)`
- `(LeftAction, RightSide)` by `(CovariantAction, RightSide)`
- `(RightAction, LeftSide)` by `(ContravariantAction, LeftSide)`
- `(RightAction, RightSide)` by `(ContravariantAction, RightSide)`

This way we don't even need `Forward`/`Backward`, which are now just `LeftSide` and `RightSide`.

**Comment by olivierverdier** [comment]
> Yes I do not follow any of this. As I said, I am okay with just being the stupid coding monkey of this. There was already a reason why I stopped trying to follow all the 0.10.0 rework, both because I felt it was redundant as well as a lot of statement that everything is fundamentally wrong and should definetly be changed, but I never got around to what is wrong.

But you are right, it was not quite as it should. It was also my intuition.

Great if we can fix all this with a breaking release.

**Comment by olivierverdier** [comment]
> Ah I see in the docs: Left backward does a multiplication from the right. Very intuitive.

Yes, this is what is fixed with the new naming, as Left backward (an alias for LeftAction, RightSide) would become CovariantAction, RightSide, which is much more intuitive indeed!

**Comment by kellertuer** [comment]
There. Is. No. Breaking. Release.

this is a completely new package. this does not break anything.
Breaking is only when we remove the old code from Manifolds.jl, but that is, from what I see now, far in the future, if we even spent that many comments on 2.5 names, we finish this rework in just a few years.

Hm. Now I am even more confused. Now Sides are also involved? But you do not change them, so why are they involved?

**Comment by olivierverdier** [comment]
Ok, no breaking release, but let's face it: people will have to adapt from `Manifolds.jl` to `LieGroups.jl`. I call that breaking even if it is not.

**Comment by olivierverdier** [comment]
> Hm. Now I am even more confused. Now Sides are also involved? But you do not change them, so why are they involved?

It's a simpler solution. One less thing to change. Call it LeftSide, Left, LeftMultiplication, whatever you prefer. It is currently LeftSide, and it's a decent name, so why not?

**Comment by kellertuer** [comment]
I would call that a “If you are coming from Manifolds.jl” documentation page here.
And sure they have to change their code, but it is not a breaking release of Manifolds per se that we introduce this.

Well let's see. For now I try to find about an hour a day to work on this,  let's see how long I can keep that and how far that gets us. Today at least: 0 knowledge on my side, 0 code lines, some brain cells lost.

**Comment by kellertuer** [comment]
> It's a simpler solution. One less thing to change. Call it LeftSide, Left, LeftMultiplication, whatever you prefer. It is currently LeftSide, and it's a decent name, so why not?

Oh mainly LeftSide was not mentioned here before nor is it changed in your comment, so why does it suddenly appear? Sure we can keep it, I just do not understand why it suddenly appears (I also have not understood what it is for, but that part with me an understanding we already discussed enough now).

**Comment by kellertuer** [comment]
(Note that we already have a side https://juliamanifolds.github.io/Manifolds.jl/stable/manifolds/group.html#Manifolds.RightSide)

**Comment by olivierverdier** [comment]
Of course, keeping Left/RightSide, it's the simplest thing to do.

**Comment by kellertuer** [comment]
but now instead of 4 types you propose 2 types and tuples of them? We should maybe write final conclusion here. And since you lost me about half an hour ago – maybe just propose what to use for the last argument of translate and I do that tomorrow

**Comment by kellertuer** [comment]
also: I started with translation, which to me is only group operations, but you also jumped over to actions now? So are we with group operations or with group actions? Somehow I feel this thread jumped back and forth between those two as well all the time, not sure why.

**Comment by kellertuer** [comment]
(and yes I probably need a book Lie groups for Dummies, I am aware of that, but your random shuffling really killed my brain for today, I still have no clue what the shuffling was for nor did, I just accept the final result).

**Comment by kellertuer** [comment]
So to conclude (sorry for such a lengthy thing here:

> So you should replace everywhere `LeftAction` by `CovariantAction` and `RightAction` by `ContravariantAction`.
> 
> In particular:
> 
> * `(LeftAction, LeftSide)`  by `(CovariantAction, LeftSide)`
> * `(LeftAction, RightSide)` by `(CovariantAction, RightSide)`
> * `(RightAction, LeftSide)` by `(ContravariantAction, LeftSide)`
> * `(RightAction, RightSide)` by `(ContravariantAction, RightSide)`
> 
> This way we don't even need `Forward`/`Backward`, which are now just `LeftSide` and `RightSide`.

And probably also remove the `ActionAndSide` type and probably just have a tuple.

**Comment by olivierverdier** [comment]
Yes, good conclusion. 👍

**Comment by kellertuer** [comment]
I am stuck on a next thing in the plethora of functions we have defined in Manifolds.jl.

We have
* apply (https://juliamanifolds.github.io/Manifolds.jl/stable/features/group_actions.html#Manifolds.apply-Tuple{AbstractGroupAction,%20Any,%20Any}) denoted by $\tau_a$ and for example translation action to be used here. In the end it is composing some a with p (applying the group operation to a manifold point, but the manifold point is not p but a – just for confusing some people)
* translate, also denoted by $\tau_q$ 

both have an inverse, both have diffs, both inverses have diffs. Now for translate we remove the in verses since we have the magic we discussed here. The only difference I can see for now is that one uses a the other q in notation. And that apply does not have all the magic with left/right/contra/co/withershinss/....

Besides that both always speak of actions and directions and such. The only difference is apply acts on manifolds and translate only maybe? For translate I could actually not find out what q is. but if q is a group element, then the only difference is that translate is basically apply with M=G?

Reverse engineering other peoples code (ok to some extend my own) that is only mildly well documented is quite time consuming. Maybe just doing it anew and seeing where this then fails is the easiest approach.

**Comment by kellertuer** [comment]
Ah then there is compose which also is just pq. But probably I am really just stupid again that I thing pq = pq = pa (if M=G).

**Comment by olivierverdier** [comment]
You are absolutely right, there are too many functions doing the same thing.

**Comment by kellertuer** [comment]
Phew! I already thought again I missed something. Maybe it grew somehow into that. For now I just added compose here then (and played a bit with a Logo idea). I will check literature for good names.
`apply` and `act` (from action) sound reasonable to me.

**Comment by olivierverdier** [comment]
I think that perhaps `translate` and `translate_diff` should be deprecated in favor of `apply`, `apply_diff`, and `apply_diff_group`.

Question: are these two calls (which produce the same result) as efficient?
```julia
translate(G, p, q)
```
and 
```julia
apply(GroupOperationAction(G), p, q)
```

**Comment by kellertuer** [comment]
Yes, they are of same efficiency. `translate` would just “check” the operation specified.

 What would be the difference between `apply_diff` and `apply_diff_group`?

**Comment by olivierverdier** [comment]
`apply` takes two arguments, a group element `g` and a point `x`, you can either differentiate w.r.t. the group element `g` (`apply_diff_group`) or w.r.t. the point `x` (`apply_diff`).

**Comment by kellertuer** [comment]
Ah, sure. In a well-documented way, this should be clear, without the signature even, that was a bit short. But sure, that sounds like reasonable names.
Will add them to the package either later today or tomorrow and document them carefully.

**Comment by olivierverdier** [comment]
Perhaps `apply_diff_point` could be a better name, instead of `apply_diff`?

**Comment by olivierverdier** [comment]
Or, `apply_diff_manifold`, to use the same logic as `apply_diff_group`?

**Comment by kellertuer** [comment]
I think manifold is better, since point is ambiguous.

**Comment by olivierverdier** [comment]
I think the following identities hold no matter what:
```julia
translate(G, p, q, conv) == apply(GroupOperationAction(G, conv), p, q)
translate_diff(G, p, q, X, conv) == apply_diff(GroupOperationAction(G, conv), p, q, X)
```

**Comment by kellertuer** [comment]
at least as far as I was able to understand the code, yes.

**Comment by mateuszbaran** [comment]
I'm trying to catch up to you. I totally agree that we have too many functions doing the same thing and making a new package is a good moment to simplify the interface. It just grew over time to accommodate more and more use cases. 

> But we also have L p ′ ( q ) = p − 1 q and R p ′ = q p − 1 – (backward) translations? Why? The inverses do exactly the same thing furthermore also have the inverse backwards ones which are just the forward non inverse ones.

Because we couldn't agree which two of these are enough. So now we have all four. It's probably better to keep all four and remove the `inverse_*` functions.

> The root cause of the problem is the tradition to call "covariant" action a left action, and a "contravariant" action a right action. This is a huge mistake (done by mathematicians).

> <warning, category math ahead>
> Covariant and contravariant are not that stupid in fact: a left action is precisely a covariant functor from a group to the permutations of a set, and a right action is a covariant functor.

That's an interesting perspective I haven't encountered yet. This sounds reasonable but are there any papers that use this terminology? They are hard to find because there is lots of papers where the words "action" and "covariant" are used in a different context.

**Comment by kellertuer** [comment]
Don't worry about the length – most of it is a proposal for new names, whose derivation I do not understand (still), but I am fine with the result. Sure backing that up by a paper would be nice, but I will try to document that in detail anyways.

**Comment by mateuszbaran** [comment]
> Question: are these two calls (which produce the same result) as efficient?
> 
> ```julia
> translate(G, p, q)
> ```
> 
> and
> 
> ```julia
> apply(GroupOperationAction(G), p, q)
> ```

They should be equivalent, and efficiency is just a matter of which one has received more optimization-related attention. Neither one is inherently less efficient. `apply` is more generic so the only reason to have `translate` is to have a name people are more familiar with. `apply` sounds fairly generic. I think not having `translate` would also be fine.

> Don't worry about the length – most of it is a proposal for new names, whose derivation I do not understand (still), but I am fine with the result. Sure backing that up by a paper would be nice, but I will try to document that in detail anyways.

I see. Whatever the final design is, we definitely need to have it well documented because the topic is, as we can see, fairly difficult to follow.

**Comment by kellertuer** [comment]
I am already looking to back up
* every doc string with according math formulae
* looking for nice literature to further provide details (work-in-progress to find a good general reference)

**Comment by Affie** [comment]
Hi this caught my eye as I spent too much time trying to figure out this forward/backward left/right stuff and I'm not sure adding more definitions in will help.
So I would be in for simplifying things. I was wondering why the combinations in ActionDirectionAndSide and if it was even needed. 
What about something in the lines of:
`Lp : left_translate(M, p, q) = p∘q`
`Rp: rigth_translate(M,p,q) = q∘p`
And if the inverses are needed.
`L'p: inv_left_translate(M,p,q) = inv(M, p) ∘q`
`R'p: inv_right_translate(M,p,q) = q∘inv(M,p)`

I would even consider creating function as that is what I see Lp, Rp to be, for example
```julia
left_translate(M, p) = (q)->compose(M, p, q)
# the usage could be
Lp = left_translate(M,p)
Lp(q)
# equivalent to
left_translate(M, p, q)
```
Please take this as intended to convey an idea and forgive any mistakes or inaccuracies, but feel free to educate me if I have the concept completely wrong.

**Comment by Affie** [comment]
> Ok, no breaking release, but let's face it: people will have to adapt from Manifolds.jl to LieGroups.jl. I call that breaking even if it is not.

Yes please make it as easy as possible to upgrade, I understand some things will break, but even an `error("deprecated, please use...)` goes a long way. The v0.9 to v0.10 upgrade affects about 8 packages.

**Comment by kellertuer** [comment]
> Hi this caught my eye as I spent too much time trying to figure out this forward/backward left/right stuff and I'm not sure adding more definitions in will help.

We will not add more definitions, we choose _one_ naming to _replace_ translate/apply. Or apply will survive but with a better naming since forward/backward was our invention and is as far as I see not used otherwise.

> So I would be in for simplifying things. I was wondering why the combinations in ActionDirectionAndSide and if it was even needed. What about something in the lines of: `Lp : translate_left(M, p, q) = p∘q` `Rp: translate_right(M,p,q) = q∘p` And if the inverses are needed. `L⁻¹p: inv_translate_left(M,p,q) = inv(M, p) ∘q` `R⁻¹p: inv_translate_right(M,p,q) = q∘inv(M,p)`

Sure one can do 4 definitions, but the idea of having that as a parameter is, that this parameter can easily be moved/changed during computations, for example in Chain rules involved in AD. With 4 functions this would always be a hustle to switch, we would work with passing function handles around and have to dispatch on `typeof(f)`.

Now one could still introduce your 4 function ideas as an “easy interface” to use, but then we have to tell everyone somehow: Ah, if you want to use your implementation for this or that or AD, please rewrite it to the other notation we have. Even worse if we _also_ introduce _further_ functions
(`apply_left(G, g, h)` and the function `LeftApplication(G,g)`).
So since you mention the chain rule – and you probably want to use it there then, we need the indication of left/right and p-or-p-inverse-whatever-its-called as parameters.

Note that this is about _actions_, not about _composition_ (group operations). `compose` stays as is).
Yes actions are a bit more general and have composition as its social case. But since I am also learning this along the way more precisely, I will also add formulae to every doc string (which is for now not yet the case in Manifold.jl for that area), cf. for example https://juliamanifolds.github.io/LieGroups.jl/previews/PR2/interface/#LieGroups.LieGroup (Work in progress, alsready saw a typo).

For the transition I will open a new issue to keep track of that and answer there.

**Comment by olivierverdier** [comment]
> > Ok, no breaking release, but let's face it: people will have to adapt from Manifolds.jl to LieGroups.jl. I call that breaking even if it is not.
> 
> Yes please make it as easy as possible to upgrade, I understand some things will break, but even an `error("deprecated, please use...)` goes a long way. The v0.9 to v0.10 upgrade affects about 8 packages.

I agree. In fact, everything discussed so far can be done without breaking anything. I would in fact suggest to apply these changes in Manifolds.jl with deprecation warnings, something like:
```julia
@deprecate LeftAction CovariantAction # or something else
@deprecate RightAction ContravariantAction # or something else
@deprecate translate(G, p, q, conv) apply(GroupOperationAction(G, conv), p, q)
@deprecate translate_diff(G, p, q, X) apply_diff(GroupOperationAction(G, conv), p, q, X)
@deprecate LeftForwardAction() (CovariantAction(), LeftSide())
# etc.
```
Then you can wait as much as *two* breaking releases before you remove the deprecated elements (this is for instance the [numpy deprecation policy](https://numpy.org/neps/nep-0023-backwards-compatibility.html)), so users have plenty of time to adapt.

**Comment by kellertuer** [comment]
Deprecation does – as far as I know – now work between packages. But we can add manual depreciation messages to all function in Manifolds.jl, what they become here.
And waiting 2 breaking releases is a nice idea, but we do so few breaking releases, that that would by now more be like 1-2 years of always keeping a lot of old code base.

And it is of course a lot of work to add all these deprecation warning, but if someone wants to do this. I will work on such things only after I finished everything here. I suspect that will not before Christmas and only if we are lucky before easter. So then this package will only go live – after that so in a year? Sure. We can do that.

And comparisons to Python I do not like much, whenever I use python, some code broke without any notice, mainly bumpy or scipy or a switch from a minor version of python to the next.

**Comment by kellertuer** [comment]
Also compare to lumpy (we got compared to them once already) – they not only are basically a company, they also have a bit more users than we have. Compared to them I would consider myself a complete amateur – I only do all this software development basically in my free time (or by writing a few less papers). So maybe that comparison is not fair.

**Comment by kellertuer** [comment]
Well, hehe. That is one of the reasons I finally want to understand the code base better and rewrite it, the old sometimes does not clearly distinguish whether its composition (or group operation) or action. Note that the action _act_ on a manifold, so we have $G\times \mathcal M$ there (but you can plug in G as M and get composition).
So translation was maybe compose, maybe action, by now I am even not so sure myself

So to not confuse you too much – `compose(G,g,h)` will stay as is (and probably what you call operation) and `apply(A, g, p, dir)` will do the group action – I will also more consistently use g and h for group elements and p,q for manifold elements, because I feel the old code was not too precise there (who ever wrote that, part of that probably also me).
If that is easier to use we might also do `adjoint_apply` or so as a function, I am not yet far enough in reading to decide that yet, but feel free to propose ideas.

**Comment by kellertuer** [comment]
I think I found a good idea what to replace translate and compose with. Translate had the problem that it (to me) was not so clear which case is which and there were 4 and 4 for the inverse – each doubled – and that translate....and all are just a certain compose.

So I will change this, which will also make the `compose_diff_left` easier to read, take the derivative of the compose with respect to the left argument.

Sure if we reintroduce the `GroupOperationAction` which turns `apply(GOA,...)` into compose, we also have the “high-level-access” to the compose functions I just sketched and can do all fancy AD rules. Maybe that is a good tradeoff to have both easy to read low-level things (`compose`) and a high-level AD-ready (like in HD ready monitors) things (`apply`)

**Comment by kellertuer** [comment]
I think I have one further question for this infinite thread.

Currently in Manifolds.jl the group action consists of a group, a manifold and a direction. But apply also has a direction (and withershins/deosil) – why is this double? Or does it make sense to remove the direction from the group action then? Or form apply and just keep withershins/deosil (sorry still have to understand the actual naming).

**Comment by Affie** [comment]
I see there are 2 ways to translate with different documentation:
[translate](https://github.com/JuliaManifolds/Manifolds.jl/blob/e31e7b088b402af19741ef3b300eccc850d96edc/src/groups/group.jl#L836-L850)
    translate(G::AbstractDecoratorManifold, p, q, conv::ActionDirectionAndSide=LeftForwardAction()])

Translate group element ``q`` by ``p`` with the translation ``τ_p`` with the specified
`conv`ention, either left forward (``L_p``), left backward (``R'_p``), right backward (``R_p``)
or right forward (``L'_p``), defined as
```math
\begin{aligned}
L_p &: q ↦ p \circ q\\
L'_p &: q ↦ p^{-1} \circ q\\
R_p &: q ↦ q \circ p\\
R'_p &: q ↦ q \circ p^{-1}.
\end{aligned}
```

and apply with the [GroupOpperationAction](https://github.com/JuliaManifolds/Manifolds.jl/blob/e31e7b088b402af19741ef3b300eccc850d96edc/src/groups/group_operation_action.jl#L1-L14)
    GroupOperationAction{AD<:ActionDirection,AS<:GroupActionSide,G<:AbstractDecoratorManifold} <: AbstractGroupAction{AD}

Action of a group upon itself via left or right translation, either from left or right side.
An element `p` of the group can act upon another another element by either:
* left action from the left side: ``L_p: q ↦ p \circ q``,
* right action from the left side: ``L'_p: q ↦ p^{-1} \circ q``,
* right action from the right side: ``R_p: q ↦ q \circ p``,
* left action from the right side: ``R'_p: q ↦ q \circ p^{-1}``.

**Comment by kellertuer** [comment]
I did not mean the translation – that would be the GroupOperationAction by now – but for cases where we fix the action, we now have the direction twice.
`apply(GroupAction, a, p, direction)`, where the group action also contains a direction already (currently / in Manifolds.jl). That is double.

**Comment by kellertuer** [comment]
continued in #6
